### DISCUSSION TITLE
InputField dataTest prop doesn't target the input itself

### DISCUSSION THREAD

**Discussion by martinkrulltott** [dhis2/notes#218] [Q&A] [answered]
When using an `InputField` with the `dataTest` prop like this:
`<InputField value={null} onChange={() => {}} dataTest={'mycustomname'}   />`
the data-test prop doesn't target the input itself, as seen in the outputted HTML: 

><div data-test="mycustomname" class="jsx-1141087211 ">
           <div data-test="mycustomname-content" class="jsx-3375842594 ">
                <div data-test="dhis2-uicore-box" class="jsx-1220078726 ">
                     <div data-test="dhis2-uicore-input" class="jsx-3353877153 jsx-1090541828 input">
                          <input type="text" class="jsx-3353877153 jsx-1090541828 " value="">
                          <div class="jsx-3353877153 jsx-1090541828 status-icon">
                       </div>
                   </div>
              </div>
         </div>
    </div>

The `mycustomname` only makes it down 2 levels and is then replaced by `dhis2-uicore-box` and `dhis2-uicore-input`, while the `input` itself doesn't get a `data-test` prop at all.

If the purpose is to be able to target an element when testing, I suggest that the `data-test` propagates down all the way to the `input`, which is really what's being tested here.

**Comment by Mohammer5** [comment, marked as answer]
The currently applied strategy is to only add a `data-test` prop to the most outer dom element of each component.
The `InputField` uses two components, so there will be three dom elements with a `data-test` attribute (the `InputField`s, `Box`es and `Input`'s outer containers). We only pass down the `dataTest` value internally if a component consists of multiple components (like the Transfer component for example).

I'd suggest to just use `[data-test="mycustomname-content"] input`. That will make the code a bit more resilient to change as the internals and/or data-test strings could change, but the element you're trying to target is probably still an `input`

**Reply by ghost** [reply]
This is not how I understand our current strategy. As I understand it we don't restrict the data-test attributes we pass down to apply just to components. I think targeting the element directly is exactly what are trying to prevent with the data-test attributes.

The goal of the data-test attributes is to allow users of our lib to target elements reliably in their tests, as a fallback for when aria attributes (or other preferred selectors) aren't sufficient. Targeting an input element in our Input components sounds like a good usecase for a data-test attribute. If it doesn't exist yet I think we should add it.

**Reply by martinkrulltott** [reply]
I can't speak for the current strategy, but as a user of the components I'd expect the purpose of the `dataTest` to be to facilitate targeting an element reliably without needing to know the inner workings of the component itself. So it would be useful if e.g. an `Input` or `InputField` would either pass the `dataTest` down to the core element (could be hard to decide when multiple core elements are used) or (preferably) pass down the `dataTest` with an element suffix, e.g. `${dataTest}-input`, so that e.g. `[data-test="mycustomname-input"]` can be targeted directly.
However, since an `Input` will most likely always contain an `input`, using `[data-test="mycustomname-content"] input` isn't too bad either (which is why I initially marked @Mohammer5 's response above as the `Answer`)

**Reply by ghost** [reply]
> as a user of the components I'd expect the purpose of the dataTest to be to facilitate targeting an element reliably without needing to know the inner workings of the component itself.

Yes, that's my understanding of the intention there as well.

> or (preferably) pass down the dataTest with an element suffix, e.g. ${dataTest}-input

Yes, that's the strategy we currently use in ui. And if the naming can be ambiguous, then the chosen suffix should be a bit more specific. For example, like what we've done with the SplitButton, which has two buttons, where one is `${dataTest}-button` and one is `${dataTest}-toggle`: https://github.com/dhis2/ui/blob/master/packages/core/src/SplitButton/SplitButton.js#L89

**Reply by Mohammer5** [reply]
Aha, interesting, I guess we never really talked about what exactly we want to target and what not to target. @ismay maybe it makes sense to schedule a call for this and talk about it? Because as far as I can tell, we currently do not add data-test props to "core" elements but instead only add them on the most outer element of a component and only pass it down in certain cases.

For example the `DropdownButton` does not have a data-test attribute on the button itself, only on the container and passes the dataTest value down to the `Popper` component. So does the `FileInput`.

I can also find a couple of examples where we do add a data-test attribute to sub elements.

---

I guess it'd make sense to put data-test attributes on all elements that the user can interact with? I'm totally open to this idea (and it's probably a good idea), my response was based on the understanding that we don't want to add too many data-test attributes

**Reply by ghost** [reply]
So when I added them to our components, I just added what was necessary for our internal tests to pass. The rationale was that we would add them for other elements whenever we saw the need for them. So the current state of the repo is not necessarily representative of our intentions. Basically, if someone needs it, and it's a relevant part of the component (like a button, or something else that'll be a persistent part of the component which could rationally be used in a test) then we should add it. Doesn't matter if its an element, or a component (as it's not just components that the user interacts with, also the elements).

**Reply by ghost** [reply]
Let me know if you'd like to plan a call btw!

**Reply by HendrikThePendric** [reply]
I think guaranteeing the presence of a data-test attribute on the outer element(s) is as far as we should take this. On some components we may be passing down the base data-test string plus a suffix, but I definitely think we should NOT guarantee this is always the case. Cypress offers a lot of commands that can help a dev find the correct DOM element within the container with the data-test attribute.

**Reply by martinkrulltott** [reply]
> Cypress offers a lot of commands that can help a dev find the correct DOM element within the container with the data-test attribute.

My concern with this approach is that any changes to the internal DOM tree structure could become a breaking change for testing purposes for apps. Putting a `data-test` prop (with a suffix if needed) at appropriate places decouples the app tests from the internal tree.

**Reply by ghost** [reply]
> Putting a data-test prop (with a suffix if needed) at appropriate places decouples the app tests from the internal tree.

Yes, that is also the purpose as I see it. About the coupling of those data-test attributes to our versioning (which is how I read Hendrik's comment), I'm not sure. But I think Martin's definition should be leading our implementation of the data-test attributes, as that's their main (if not only) purpose.

**Reply by amcgee** [reply]
This doesn't seem to be a unique situation 
- https://github.com/dhis2/notes/discussions/219 
- https://jira.dhis2.org/browse/LIBS-118

**Reply by varl** [reply]
It's not, and it's a very good discussion to have now that UI components are being used in various apps. It's not necessarily so that we got the `data-test` strategy right on the first (or second) try, and the input (no pun intended) here is very valuable to consider.

We don't want to tie tests to internal structure for the application tests, however that's fine for tests in UI. We have a consensus and commitment that breaking changes means that the public facing API has changed, but we are free to change the internals of UI components however we want. We cannot use the default recommendation then that App developers target UI component internals for testing purposes.

Obviously this bites of a large piece that we need to chew on for a bit, so for the purposes of what the answer to this question is today is to use: `[data-test="mycustomname-content"] input ` for now while we work on revisiting how `dataTest` is implemented.

**Comment by amcgee** [comment]
I created a proposal to formalize our `dataTest` concept and approach, feedback welcome:

https://github.com/dhis2/notes/discussions/224
